lib/internal/modules/cjs/loader.js

```diff
@@ -728,7 +729,11 @@ if (experimentalModules) {
 // bootstrap main module.
 Module.runMain = function() {
   // Load the main module--the command line argument.
-  if (experimentalModules) {
+  const base = path.basename(process.argv[1]);
+  const ext = path.extname(base);
+  const isESM = ext === '.mjs';
+
+  if (experimentalModules && (isESM || hasLoader)) {
```

Comment:
I would prefer if we could move this filtering logic into the resolver itself still, to ensure a single code path for resolution.